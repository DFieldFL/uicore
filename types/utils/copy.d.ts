import copy from 'clipboard-copy';
export default copy;
