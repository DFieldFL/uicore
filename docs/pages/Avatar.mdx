# Avatar --- docs/pages/Avatar.mdx


## Import

```js
import { Avatar } from '@wings-software/uikit'
```


```js live=true
<Layout.Horizontal spacing="small" id="primary-buttons">
  <Avatar name="John Doe"/>
  <Avatar email="john.doe@harnes.io"/>
</Layout.Horizontal>
```