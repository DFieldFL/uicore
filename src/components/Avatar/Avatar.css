.Avatar {
  display: inline-block;
  margin: 0 5px;
}

.AvatarInner {
  font-weight: bolder;
  box-sizing: border-box;
  white-space: nowrap;
  text-overflow: ellipsis;
  color: white;
}
