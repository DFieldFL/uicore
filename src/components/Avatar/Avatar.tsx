import React, { useEffect, useState } from 'react'
import css from './Avatar.css'

const defaultColors = ['#eb2f06', '#e58e26', '#1e3799', '#78e08f', '#079992', '#ff6b81', '#a4b0be']

export interface AvatarProps {
  borderRadius?: number
  name?: string
  src?: string
  color?: string
  colors?: string[]
  size?: number
  style?: React.CSSProperties
  className?: string
  email?: string
}

export const Avatar = (props: AvatarProps) => {
  const {
    borderRadius = '100%',
    src,
    name = '',
    color,
    colors = defaultColors,
    size = 36,
    style,
    className,
    email = ''
  } = props

  const [backgroundColor, setBackgroundColor] = useState('')

  useEffect(() => {
    if (color) {
      setBackgroundColor(color)
    } else {
      const i = Math.floor(Math.random() * 10) % colors.length
      setBackgroundColor(colors[i])
    }
  }, [])

  const initialsFromName = name
    .split(/-| /)
    .map((n: string) => n[0])
    .join('')
  const initialsFromEmail = email
    .split('.')
    .map(n => n[0])
    .join('')
    .slice(0, 2)
    .toUpperCase()

  const formatedSize = String(size) + 'px'

  const imageStyle: React.CSSProperties = {
    display: 'block',
    borderRadius
  }

  const contentStyle: React.CSSProperties = {
    lineHeight: formatedSize,
    textAlign: 'center',
    borderRadius
  }

  if (formatedSize) {
    imageStyle.width = contentStyle.width = contentStyle.maxWidth = formatedSize
    imageStyle.height = contentStyle.height = contentStyle.maxHeight = formatedSize
  }

  let inner
  const classes = [className, css.Avatar]
  if (src) {
    inner = <img style={imageStyle} src={src} alt={name} />
  } else {
    contentStyle.backgroundColor = backgroundColor
    inner = initialsFromName || initialsFromEmail || 'X'
  }

  return (
    <div className={classes.join(' ')} style={style}>
      <div className={css.AvatarInner} style={contentStyle}>
        {inner}
      </div>
    </div>
  )
}
