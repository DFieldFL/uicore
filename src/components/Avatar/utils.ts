export const getInitialsFromNameOrEmail = (name = '', email = '') => {
  let initialsFromName = name
    .split(/-| /)
    .map((n: string) => n[0])
    .join('')

  initialsFromName =
    initialsFromName.length > 2
      ? `${initialsFromName[0]}${initialsFromName[initialsFromName.length - 1]}`
      : initialsFromName
  let initialsFromEmail = email.split('@')[0]
  const splitedInitialsFromEmail = initialsFromEmail.split('.')
  initialsFromEmail =
    splitedInitialsFromEmail.length > 1
      ? `${splitedInitialsFromEmail[0][0]}${splitedInitialsFromEmail[splitedInitialsFromEmail.length - 1][0]}`
      : initialsFromEmail[0]
  return initialsFromName || initialsFromEmail
}
