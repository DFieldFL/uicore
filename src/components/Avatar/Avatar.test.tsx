import React from 'react'
import { render } from '@testing-library/react'
import { Avatar } from './Avatar'
import { getInitialsFromNameOrEmail } from './utils'

describe('Render basic component', () => {
  test('should check snapshot with name', () => {
    const { container } = render(<Avatar backgroundColor="red" name="John Doe" />)
    expect(container).toMatchSnapshot()
  })
  test('should check snapshot with email', () => {
    const { container } = render(<Avatar backgroundColor="red" email="John.Doe@harness.io" />)
    expect(container).toMatchSnapshot()
  })
  test('should check snapshot with image', () => {
    const { container } = render(
      <Avatar backgroundColor="red" src="https://www.mangaship.com/Content/img/human-avatar.png" />
    )
    expect(container).toMatchSnapshot()
  })
  test('getInitials with name', () => {
    expect(getInitialsFromNameOrEmail('All Name Here')).toEqual('AH')
    expect(getInitialsFromNameOrEmail('All Name')).toEqual('AN')
    expect(getInitialsFromNameOrEmail('AllNameHere')).toEqual('A')
  })
  test('getInitials with email', () => {
    expect(getInitialsFromNameOrEmail('', 'AllNameHere@com')).toEqual('A')
    expect(getInitialsFromNameOrEmail('', 'AllName.here@com')).toEqual('Ah')
  })
})
