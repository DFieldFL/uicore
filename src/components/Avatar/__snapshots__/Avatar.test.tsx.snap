// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render basic component should check snapshot with email 1`] = `
<div>
  <div
    class="Avatar contentStyle"
  >
    <div
      class="AvatarInner"
      style="border-radius: 100%; line-height: 36px; width: 36px; height: 36px; background-color: red;"
    >
      JD
    </div>
  </div>
</div>
`;

exports[`Render basic component should check snapshot with image 1`] = `
<div>
  <div
    class="Avatar contentStyle"
  >
    <div
      class="AvatarInner"
      style="border-radius: 100%; line-height: 36px; width: 36px; height: 36px;"
    >
      <img
        alt=""
        class="imageStyle"
        src="https://www.mangaship.com/Content/img/human-avatar.png"
        style="border-radius: 100%; line-height: 36px; width: 36px; height: 36px;"
      />
    </div>
  </div>
</div>
`;

exports[`Render basic component should check snapshot with name 1`] = `
<div>
  <div
    class="Avatar contentStyle"
  >
    <div
      class="AvatarInner"
      style="border-radius: 100%; line-height: 36px; width: 36px; height: 36px; background-color: red;"
    >
      JD
    </div>
  </div>
</div>
`;
